lib/promscrape/scrapework.go

```diff
@@ -1051,6 +1056,36 @@ func (wc *writeRequestCtx) addRows(cfg *ScrapeWork, rows []parser.Row, timestamp
 
 var errLabelsLimitExceeded = errors.New("label_limit exceeded")
 
+func (wc *writeRequestCtx) addMetadata(metadataRows []parser.Metadata) {
+	if len(metadataRows) == 0 {
+		return
+	}
+	uniqMetadata := make(map[string]*prompbmarshal.MetricMetadata, len(metadataRows)/2)
+
+	for i := range metadataRows {
+		row := &metadataRows[i]
+		md, ok := uniqMetadata[row.Metric]
+		if !ok {
+			md = &prompbmarshal.MetricMetadata{
```

Comment:
This function allocates memory, which I'd like to avoid. But maybe it's better to add it as is and improve performance later. I did quick benchmark and it shows x2 parsing performance degradation.